src/lib/raw_data/xml_reader.cpp

```diff
@@ -478,8 +478,14 @@ std::optional<RawData::RawData> XmlReader::read_mzml(
                 }
 
                 if (tag.value().name == "precursor") {
-                    // TODO: Get precursor id from "spectrumRef" attribute.
-                    scan.precursor_information.scan_number = 0;
+                    auto precursor_attributes = tag.value().attributes;
+                    std::string spectrumRef =
+                        precursor_attributes["spectrumRef"];
+                    // Find scan number.
+                    size_t scanIdx = spectrumRef.find("scan=") + 5;
+                    scan.precursor_information.scan_number =
+                        stoull(spectrumRef.substr(scanIdx));
```

Comment:
should be `std::stoull` I think since we are not `using std::string`. I don't like `using` in general since the namespaces help document where things are coming from.